src/main/java/com/slack/kaldb/metadata/ZookeeperMetadataStore.java

```diff
@@ -0,0 +1,337 @@
+  private final ListeningExecutorService metadataExecutorService;
+
+  public ZookeeperMetadataStore(
+      String zkHostPath,
+      String zkPathPrefix,
+      int sessionTimeoutMs,
+      int connectionTimeoutMs,
+      RetryPolicy retryPolicy,
+      FatalErrorHandler fatalErrorHandler,
+      MeterRegistry meterRegistry) {
+    ensureNonEmptyString(zkHostPath, "zkHostPath can't be null or empty");
+    ensureNonEmptyString(zkPathPrefix, "zkPathPrefix can't be null or empty");
+    ensureTrue(sessionTimeoutMs > 0, "sessionTimeoutMs should be a positive number");
+    ensureTrue(connectionTimeoutMs > 0, "connectionTimeoutMs should be a positive number");
+
+    this.failureCounter = meterRegistry.counter(METADATA_FAILED_COUNTER);
+    this.zkFailureCounter = meterRegistry.counter(ZK_FAILED_COUNTER);
+    this.metadataWriteCounter = meterRegistry.counter(METADATA_WRITE_COUNTER);
+    this.metadataReadCounter = meterRegistry.counter(METADATA_READ_COUNTER);
+
+    // TODO: Pass the thread pool in a constructor so we can use direct executor in tests if needed.
+    // Create an executor service that runs all the ZK operations
+    ThreadFactory metadataStoreThreadFactory =
+        new ThreadFactoryBuilder().setNameFormat("kaldb-metadata-store-pool").build();
+    ThreadPoolExecutor executor =
+        (ThreadPoolExecutor) Executors.newCachedThreadPool(metadataStoreThreadFactory);
+    this.metadataExecutorService =
+        MoreExecutors.listeningDecorator(
+            MoreExecutors.getExitingExecutorService(executor, 1, TimeUnit.SECONDS));
+
+    // TODO: In future add ZK auth credentials can be passed in here.
+    this.curator =
+        CuratorFrameworkFactory.builder()
+            .connectString(zkHostPath)
+            .namespace(zkPathPrefix)
+            .connectionTimeoutMs(connectionTimeoutMs)
+            .sessionTimeoutMs(sessionTimeoutMs)
+            .retryPolicy(retryPolicy)
+            .build();
+
+    // A catch-all handler for any errors we may have missed.
+    curator
+        .getUnhandledErrorListenable()
+        .addListener(
+            (message, exception) -> {
+              LOG.error("Unhandled error {} could be a possible bug", message, exception);
+              failureCounter.increment();
+              throw new RuntimeException(exception);
+            });
+
+    // Log all connection state changes to help debugging.
+    curator
+        .getConnectionStateListenable()
+        .addListener(
+            (curator, connectionState) -> {
+              LOG.info("Curator connection state changed to {}", connectionState);
+            });
+
+    /*
+     * If a ZK session expires, we need to create all the watches and ephemeral nodes again.
+     * In such a case, any ephermeral nodes and watches would expire and need to be re-created.
+     * To keep it simple for now, we terminate the process and let the process initialization
+     * register those nodes.
+     */
+    curator
```

Comment:
can you explain this a little to me ? Once we hit session expiry ( say a GC pause bigger than session timeout ) will we stop indexing ?

Reply:
I think I get it - `RuntimeHalterImpl` is what would happen - I am unsure if we want to do this rather than re-connect.

We can do the re-connect safely in a followup PR

Reply:
The general philosophy for now is to make the process crash the process for now whenever we see a mis-configuration. This is preferable at this stage than having the process continue and result in an unrelated error which would be much more hard to debug.

In this specific case, we can re-connect the session, but we need a way to re-register all the watches etc again. Since those use cases don't exist yet, I decided to keep it simple and implement re-connect and other features in a follow up PR.